pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -1365,6 +1365,12 @@ func (c *Controller) getProxyServiceInstancesByPod(pod *v1.Pod,
 func (c *Controller) GetProxyWorkloadLabels(proxy *model.Proxy) labels.Instance {
 	pod := c.pods.getPodByProxy(proxy)
 	if pod != nil {
+		locality := c.getPodLocality(pod)
+		if pod.Labels == nil {
+			pod.Labels = make(map[string]string)
+		}
+		// Add locality labels
+		pod.Labels[model.LocalityLabel] = locality
```

Comment:
@diranged it's not entirely obvious, but previously, gateway locality wasn't being respected because we derived locality from just ServiceInstances. Now, we have added a locality label to Pods (including gateways), and we use that locality with (ads.go L622):

```go
if len(proxy.ServiceInstances) > 0 {
	localityStr = proxy.ServiceInstances[0].Endpoint.Locality.Label
} else {
	// If no service instances(this maybe common for a pure client), respect LocalityLabel
	localityStr = proxy.Metadata.Labels[model.LocalityLabel]
}
``` 
The gateway case false into the "else" here.